common/eventconsumption/eventindexmodel.cpp

```diff
@@ -66,80 +116,128 @@ class CIndexEventModel : public CInterfaceOf<IEventModel>
         storage.observeFile(event.queryNumericValue(EvAttrFileId), event.queryTextValue(EvAttrPath));
     }
 
-    void onIndexCacheHit(std::vector<CEvent>& events)
+    void onIndexCacheHit(std::vector<CEvent> &events)
     {
-        CEvent& event = events.front();
+        CEvent &event = events.at(0);
         ModeledPage page;
-        bool firstObservation = expansion.observePage(event);
-        expansion.describePage(event, page);
-        if (ExpansionMode::OnLoad == page.expansionMode)
-        {
-            // A recorded hit without an index cache may transform the recorded size and time
-            // into estimated values.
-            event.setValue(EvAttrInMemorySize, page.expanded.size);
-            event.setValue(EvAttrExpandTime, page.expansionTime);
-
-            // If an index cache is in use and the page is not in the cache, the size and time
-            // transform to zeroes, and a simulated IndexLoad is required.
-        }
-        else if (ExpansionMode::OnLoadToOnDemand == page.expansionMode)
+        memory.observePage(event);
+        memory.describePage(event, page);
+        if (!page.cacheHit)
         {
-            // A recorded hit without an index cache replaces the size and time with zeroes.
-            event.setValue(EvAttrInMemorySize, page.compressed.size);
-            event.setValue(EvAttrExpandTime, 0ULL);
+            storage.useAndDescribePage(event.queryNumericValue(EvAttrFileId), event.queryNumericValue(EvAttrFileOffset), event.queryNumericValue(EvAttrNodeKind), page);
+            events.push_back(event); // copy before transforming the hit to a miss to preserve load attributes
+            CEvent &reclassified = events.at(0); // 'event' can be invalidated by the push_back
+            CEvent &simulated = events.at(1);
+            reclassified.transform(EventIndexCacheMiss);
+            simulated.transform(EventIndexLoad);
+            simulated.setValue(EvAttrEventTimestamp, simulated.queryNumericValue(EvAttrEventTimestamp) + 1ULL);
+            switch (page.expansionMode)
+            {
+            case ExpansionMode::OnLoad:
+                simulated.setValue(EvAttrInMemorySize, page.expanded.size);
+                simulated.setValue(EvAttrExpandTime, page.expansionTime);
+                simulated.setValue(EvAttrReadTime, page.readTime);
+                break;
+            case ExpansionMode::OnLoadToOnDemand:
+            case ExpansionMode::OnDemand:
+                simulated.setValue(EvAttrInMemorySize, page.compressed.size);
+                simulated.setValue(EvAttrExpandTime, 0ULL);
+                simulated.setValue(EvAttrReadTime, page.readTime);
+                break;
+            default:
+                break;
+            }
 
-            // If an index cache is in use and the page is not in the cache, the size and time
-            // transform to zeroes, and a simulated IndexLoad is required.
+            if (memory.caching(reclassified))
```

Comment:
I would be tempted to simplify the code by removing this test, but no need.  If retained, passing page.nodeKind would be more efficient .